network/test/middleware_test.go

```diff
@@ -509,33 +525,63 @@ func (m *MiddlewareTestSuite) Ping(msg *message.Message, expectID, expectedMessa
 
 }
 
-// Ping sends count-many distinct messages concurrently from the first middleware of the test suit to the last one
-// It evaluates the correctness of reception of the content of the messages, as well as the sender ID
+// MultiPing sends count-many distinct messages concurrently from the first middleware of the test suit to the last one.
+// It evaluates the correctness of reception of the content of the messages. Each message must be received by the
+// last middleware of the test suit exactly once.
 func (m *MiddlewareTestSuite) MultiPing(count int) {
 	receiveWG := sync.WaitGroup{}
 	sendWG := sync.WaitGroup{}
 	// extracts sender id based on the mock option
 	// mocks Overlay.Receive for  middleware.Overlay.Receive(*nodeID, payload)
-	firstNode := 0
-	lastNode := m.size - 1
+	firstNodeIndex := 0
+	lastNodeIndex := m.size - 1
+
+	receivedPayloads := make(map[string]struct{}) // keep track of unique payloads received.
+
+	// regex to extract the payload from the message
+	regex := regexp.MustCompile(`^hello from: \d`)
+
 	for i := 0; i < count; i++ {
 		receiveWG.Add(1)
 		sendWG.Add(1)
-		msg, expectedMsg, expectedPayload := messageutils.CreateMessage(m.T(),
-			m.ids[firstNode].NodeID,
-			m.ids[lastNode].NodeID,
+
+		expectedPayloadText := fmt.Sprintf("hello from: %d", i)
+		msg, err := network.NewOutgoingScope(
+			flow.IdentifierList{m.ids[lastNodeIndex].NodeID},
 			testChannel,
-			fmt.Sprintf("hello from: %d", i))
+			&libp2pmessage.TestMessage{
+				Text: expectedPayloadText,
+			},
+			unittest.NetworkCodec().Encode,
+			network.ProtocolTypeUnicast)
+		require.NoError(m.T(), err)
 
-		m.ov[lastNode].On("Receive", m.ids[firstNode].NodeID, expectedMsg, expectedPayload).Return(nil).Once().
+		mu := sync.Mutex{} // protects overlay.Receive
+		m.ov[lastNodeIndex].On("Receive", mockery.Anything).Return(nil).Once().
 			Run(func(args mockery.Arguments) {
-				payload := args.Get(2).(*libp2pmessage.TestMessage)
-				require.Equal(m.T(), expectedPayload.(*libp2pmessage.TestMessage), payload)
 				receiveWG.Done()
+
+				msg, ok := args[0].(*network.IncomingMessageScope)
+				require.True(m.T(), ok)
+
+				require.Equal(m.T(), testChannel, msg.ChannelId())                    // channel
+				require.Equal(m.T(), m.ids[firstNodeIndex].NodeID, msg.OriginId())    // sender id
+				require.Equal(m.T(), m.ids[lastNodeIndex].NodeID, msg.TargetIDs()[0]) // target id
+				require.Equal(m.T(), network.ProtocolTypeUnicast, msg.Protocol())     // protocol
+
+				mu.Lock() // protects receivedPayloads
```

Comment:
is this to protect concurrent access to `receivedPayloads`? You might consider using [ProtectedMap](https://github.com/onflow/flow-go/blob/master/utils/unittest/protected_map.go#L6-L6)

Reply:
https://github.com/onflow/flow-go/pull/3675/commits/350f469a64469f7d888e732cae784e6c26d5eb44